sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/EmptyNestedBlock.cs

```diff
@@ -0,0 +1,313 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.VisualBasic;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class EmptyNestedBlock : EmptyNestedBlockBase
+    {
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private readonly List<SyntaxNode> emptyInnerBlocks = new List<SyntaxNode>();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    emptyInnerBlocks.Clear();
+                    VerifyEmptyBlocks(c.Node);
+                    emptyInnerBlocks.ForEach(node => c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, node.GetLocation())));
+                },
+                SyntaxKind.SimpleDoLoopBlock,
+                SyntaxKind.DoLoopUntilBlock,
+                SyntaxKind.DoLoopWhileBlock,
+                SyntaxKind.DoUntilLoopBlock,
+                SyntaxKind.DoWhileLoopBlock,
+                SyntaxKind.ForBlock,
+                SyntaxKind.ForEachBlock,
+                // The Else and ElseIf blocks are inside the MultiLineIfBlock
+                SyntaxKind.MultiLineIfBlock,
+                SyntaxKind.SelectBlock,
+                // The CatchBlock and FinallyBlock are inside the TryBlock
+                SyntaxKind.TryBlock,
+                SyntaxKind.UsingBlock,
+                SyntaxKind.WhileBlock,
+                SyntaxKind.WithBlock);
+        }
+
+        /**
+         * Verify that the given block has no statements and no comments inside
+         * (notable exception: the Select block)
+         *
+         * Note: Roslyn maps the comments which are inside a block as trivia of the following block
+         * e.g. in the below snippet, the comment will be part of the Finally block
+         *
+         * Try
+         *   ' my comment
+         * Finally
+         * End Try
+         */
+
+        private void VerifyEmptyBlocks(SyntaxNode node)
```

Comment:
There is no need for a state in this class. This method should return `IEnumerable<SyntaxNode>` and every inner method could just `yield return` instead of `Add`.